### ISSUE TITLE
Train a Yolov5 model on different dataset and benchmark

### ISSUE THREAD

**Issue by Rhyme0730** [BUZZ-Blimps/CatchingBlimp#36]
A model's performance and robustness are vital when in competition, we don't want waste plenty of time on training a model. So it is important to train our model quickly while maintain accuracy. Here are the steps for training and testing.

1. Train a custom yolov5 model on PC using different input size (960p)
2. Convert the best model of each different training set into .rknn form to implement on Orangepi ( It may lose some accuracy to maintain model size when converting)
3. Fine tune this yolov5 model and test it in lab.
4. Compare it with Yolov8 in accuracy, fps (on board)

**Comment by Rhyme0730** [comment]
Training a yolov5 model based on a dataset of 150 images, it may cause overfitting when implementing on camera
![image](https://github.com/SWAMP-Blimps/CatchingBlimp/assets/124558489/0bcb9b42-1002-4cd4-b9bd-408e6c1f3937)

**Comment by Rhyme0730** [comment]
(TBD) Problem1: When exporting our rknn model input size as 960x1280 (h,w), it can't be directly implemented on our previous code on board. There is something to be revised in `process` function in func.py

**Comment by Rhyme0730** [comment]
Instead implementing model with input size as 960x960, it works but its fps is reduced to 16 and three npu loads all get to their highest percentage 84%, 79%, 75%, respectively. I guess larger size video input will cause video jamming. While using input size 640x640 video performs well (fps reach 30, camera's highest fps)

**Comment by sah4jpatel** [comment]
Is this a quantized model that’s yielding 16fps?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by Rhyme0730** [comment]
Yes, the actual model size is around 8.4 MB.

**Comment by Rhyme0730** [comment]
Test size 736x736, the fps can reach the highest.
Test size 960x960 again: the fps also can reach the highest and keep it about 1 min. 3 NPU core load around 50%.
But after 1 min, the fps reduce to around 16 which is the same as yesterday.
Guess the performance of NPU may be strongly influenced by the high temperature.
![image](https://github.com/SWAMP-Blimps/CatchingBlimp/assets/124558489/215f1328-1f89-4556-ab57-eee8848e0349)

**Comment by Rhyme0730** [comment]
The size 736x736's model also has some problem when the chip's temp is high, but not that obvious.

**Comment by Rhyme0730** [comment]
Currently the most stable model input is 640x640 (even the chip's temp is high). Next step is training a good model based on our previous video.

**Comment by Rhyme0730** [comment]
The model trained from video images performs well in training (see pictures below) but is not good in real implementation (It takes around an hour to train). Here are 2 possible ways to improve it.
1. Add some irrelevant images into train dataset to reduce the risk of overfitting
2. Record videos using camera in real world with different conditions (lights, dark, etc.), and then split them into new dataset.
Method 2 can be tried in later if 1 is not that good. And it is also the methods we need to implement in competition.
![image](https://github.com/SWAMP-Blimps/CatchingBlimp/assets/124558489/0cd94a62-7806-433a-8fe4-ef32a54612b5)
![image](https://github.com/SWAMP-Blimps/CatchingBlimp/assets/124558489/7b2a0843-a332-424e-95d9-57b2f0e2270b)

**Comment by Rhyme0730** [comment]
Btw, I didn't use Deepsort (but I saw the code in previous ML pipeline). I think it is also a good way to improve tracking performance, since Yolov5 always loses the tracking target when I move the camera

**Comment by Rhyme0730** [comment]
I have rewritten new ML pipeline based on RK3588 and yolov5. The whole pipeline is recording videos -> create dataset -> preprocess dataset -> Train a yolov5 model -> convert into rknn model -> Implement on orangePi. Additional shell scripts will be  finished after selecting the yolo model version

**Comment by Rhyme0730** [comment]
Trained a model based on some new dataset I recorded in Lab. It's generalization ability is not that good, which means it can't handle the scenario that didn't happen in the training dataset.
![image](https://github.com/user-attachments/assets/3d5863e9-cd68-4938-bcc7-7e1930cb9d0a)

**Comment by Rhyme0730** [comment]
![image](https://github.com/user-attachments/assets/c1d57c3b-8868-4c38-8fcf-0e94efa45245)
![image](https://github.com/user-attachments/assets/f10261a8-e26a-4129-ae92-73bc350835f1)
![image](https://github.com/user-attachments/assets/9e79c259-dc49-4eeb-8145-dc8c35846970)

I trained a good model with a 500 pictures dataset and only took 10 mins to train. The performance is good in both lab and conference room. Here are some some thoughts.
1. Yolov5s has the problem of overfitting, so it can't distinguish similar things (like a green game ball and a green book) if these things are not in dataset. In short, we need to feed yolov5s model as much different scenario's pictures as possible.
2. I think the augmentation of pictures is non-necessary because it won't improve much accruacy but leads to a low training speed. As the same as thought 1, I suggest we can record video as much as possible to select 500-1000 representive pictures.
What do you think? @sah4jpatel

**Comment by sah4jpatel** [comment]
Is this issue good to close? @Rhyme0730
